{% extends 'content.html' %}
{% load i18n %}

{% block add_head %}
{% endblock add_head %}

{% block add_foot %}

<script>

	function validateSequences() {
		var seqs_field = document.getElementById('id_sequences_raw');
		var raw_seqs = seqs_field.value;
		var raw_seqs_list = raw_seqs.split(/\r?\n/);

		// check if unallowed characters
		var chars_valid = !(new RegExp("[^ACGT]").test(raw_seqs_list.join('')));

		// check for number of sequences
		var number_valid = !(raw_seqs_list.length > 10000);

		// check for length of sequences
		var lgth = 0;
		for (var i = 0; i < raw_seqs_list.length; i++) {
			if (raw_seqs_list[i].length > lgth) {
				var lgth = raw_seqs_list[i].length;
			}
		}
		var length_valid = !(lgth > 150);

		if (chars_valid & number_valid & length_valid) {
			seqs_field.classList.remove("is-invalid")
			return true;
		} else {
			seqs_field.classList.add("is-invalid")
			seqs_field.classList.remove("is-valid")
			return false;
		}
	}


	function updateSeqStats() {
		try {
			var seqs_field = document.getElementById('id_sequences_raw');
			var raw_seqs = seqs_field.value;
			var raw_seqs_list = raw_seqs.split(/\r?\n/);

			// GC content
			var GC_content = (raw_seqs_list.join()).match(/[GC]/g).length/raw_seqs_list.join().length;

			// number of sequences
			var n_seqs = raw_seqs_list.length;

			// length of sequences
			var lgth = 0;
			for (var i = 0; i < raw_seqs_list.length; i++) {
				if (raw_seqs_list[i].length > lgth) {
					var lgth = raw_seqs_list[i].length;
				}
			}
		} catch (error) {
			var GC_content = 0;
			var n_seqs = 0;
			var lgth = 0;
		};
		
		document.getElementById('sequences-GC').textContent = Math.round(100*GC_content) + "%";
		document.getElementById('sequences-count').textContent = n_seqs;
		document.getElementById('sequences-length').textContent = lgth;
	}

	

	function clearSeqs() {
		document.getElementById('id_sequences_raw').value = "";
		updateSeqStats();
	}

	function randomSeqs() {
		var results = [];
		for (let j = 0; j < 100; j++) {
			var result = '';
			var characters = 'ACGT';
			for (let i = 0; i < 100; i++) {
				result += characters.charAt(Math.floor(4*Math.random()));
			}
			results.push(result)
		}
		document.getElementById('id_sequences_raw').value = results.join("\n");
		updateSeqStats();
	}



	function updateErrorStats() {
		var subs = 0;
		var dels = 0;
		var ins = 0;
		var bias = 0;

		var challenge = document.getElementById('id_challenge').value;

		switch(challenge) {
			case "PS":
				subs = 0.025;
				dels = 0.075;
				ins = 0.012;
				bias = 0.44;
				break;
			case "DD":
				subs = 0.0049;
				dels = 0.0007;
				bias = 0.30;
				break;
			default:
				subs = 0.0;
				dels = 0.0;
				bias = 0.0;
			}
		
		document.getElementById('errorprofile-substitutions').textContent = (100*subs).toFixed(1) + "%";
		document.getElementById('errorprofile-deletions').textContent = (100*dels).toFixed(1) + "%";
		document.getElementById('errorprofile-insertions').textContent = (100*ins).toFixed(1) + "%";
		document.getElementById('errorprofile-bias').textContent = bias.toFixed(2);
	}

	function attachToClass( classname, function_call ) {
		var selection = document.getElementsByClassName(classname);
		for(var i = 0; i < selection.length; i++) {
			(function(index) {
				selection[index].addEventListener("input", function_call)
			})(i);
			}
	}

	updateErrorStats();
	updateSeqStats()
	attachToClass('form-control', function() { updateErrorStats(); })
	attachToClass('form-select', function() { updateErrorStats(); })
	attachToClass('form-range', function() { updateErrorStats(); })
	attachToClass('form-check-input', function() { updateErrorStats(); })

	document.getElementById('id_sequences_raw').addEventListener('input', function() { validateSequences(); updateSeqStats(); });


	// Example starter JavaScript for disabling form submissions if there are invalid fields
	(function() {
	  'use strict';
	  window.addEventListener('load', function() {
		// Fetch all the forms we want to apply custom Bootstrap validation styles to
		var forms = document.getElementsByClassName('needs-validation');
		// Loop over them and prevent submission
		var validation = Array.prototype.filter.call(forms, function(form) {
		  form.addEventListener('submit', function(event) {
			if (!validateSequences() || !form.checkValidity()) {
				event.preventDefault();
				event.stopPropagation();
			}
			form.classList.add('was-validated');
		  }, false);
		});
	  }, false);
	})();
</script>

{% endblock add_foot %}

{% block content_body %}

{% load widget_tweaks %}

<form class="needs-validation" action="{% url 'jobs:challenge' %}" method="post" novalidate>

	{% csrf_token %}

	<div class="row g-3">

		<div class="col-sm-12 pb-2">
			<label for="{{ form.name.id_for_label }}" class="form-label">Job name</label>
			<div class="input-group has-validation">
				{% if form.name.errors %}
					{{ form.name|add_class:"form-control is-invalid" }}
				{% else %}
					{{ form.name|add_class:"form-control" }}
				{% endif %}
				<div class="invalid-feedback">
					A valid submission name is required. {{ form.name.errors }}
				</div>
			</div>
			<small class="form-text text-muted">
				Choose a name for your submission.
			</small>
		</div>

		

		<div class="col-sm-9 pb-2">
			<label for="{{ form.sequences_raw.id_for_label }}" class="form-label">Design sequences</label>
			<div class="input-group has-validation">
				{% if form.sequences_raw.errors %}
					{{ form.sequences_raw|add_class:"form-control is-invalid font-monospace"|attr:"placeholder:Paste your oligo sequences here" }}
				{% else %}
					{{ form.sequences_raw|add_class:"form-control font-monospace"|attr:"placeholder:Paste your oligo sequences here" }}
				{% endif %}
				<div class="invalid-feedback">
					Please provide a valid set of oligo sequences. They may only contain A, C, G, and T. At most 10000 sequences with up to 150 nucleotides each are allowed.
				</div>
			</div>
		</div>
		
		<div class="col-sm-3 pb-2 align-self-end">
			<table class="table table-borderless table-sm">
				<tbody>
					<tr>
						<td class="text-end" style="padding: 0;"><small class="text-muted">#oligos</small></td>
						<td class="text-end" style="padding: 0;"><small class="text-muted" id="sequences-count">0</small></td>
					</tr>
					<tr>
						<td class="text-end" style="padding: 0;"><small class="text-muted">max. length</small></td>
						<td class="text-end" style="padding: 0;"><small class="text-muted" id="sequences-length">0</small></td>
					</tr>
					<tr>
						<td class="text-end" style="padding: 0;"><small class="text-muted">avg. GC</small></td>
						<td class="text-end" style="padding: 0;"><small class="text-muted" id="sequences-GC">0%</small></td>
					</tr>
				</tbody>
			</table>
			
			<div class="w-100 btn btn-outline-secondary btn-sm" onclick="randomSeqs();">Random</div>
			<div class="w-100 btn btn-outline-secondary btn-sm mt-2" onclick="clearSeqs();">Clear</div>
		</div>
		
		<small class="form-text text-muted">
			Provide the oligo sequences for simulation. Enter each oligo on its own line. Limited to 10000 oligos of max. 150 nt each.
		</small>

	</div>

	<div class="row g-5 pt-5">
		<div class="col-md-7 col-lg-8">
			<div class="row g-3">

				<div class="col-12 pb-3">
					<label for="{{ form.challenge.id_for_label }}" class="form-label">Challenge</label>
					<div class="input-group has-validation">
						{{ form.challenge|add_class:"form-select" }}
						<div class="invalid-feedback">
							A valid challenge is required. {{ form.challenge.errors }}
						</div>
					</div>
					<small class="form-text text-muted">
						Select the challenge for simulation. More details below.
					</small>
				</div>	

				<label for="accordionFlushExample" class="form-label">Challenge definitions</label>
				<div class="accordion mt-0" id="accordionFlushExample">
					<div class="accordion-item">
					  <h2 class="accordion-header" id="flush-headingOne">
						<button class="accordion-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseOne" aria-expanded="false" aria-controls="flush-collapseOne">
						  DNA decay after long-term storage
						</button>
					  </h2>
					  <div id="flush-collapseOne" class="accordion-collapse collapse" aria-labelledby="flush-headingOne" data-bs-parent="#accordionFlushExample">
						<div class="accordion-body">
							Recovery of oligo fragments after long-term storage, assuming the very low error rates from state-of-the-art commercial synthesis and high-fidelity PCR, but at very low physical coverage and oligo breakage equivalent to around five half-lives of storage. In addition, the sequencing depth is low, and the sequencing data is biased against short oligo fragments.
							<ul class="pt-3">
								<li>0.0007 deletions per nt, 0.0049 substitutions per nt, and no insertions</li>
								<li>0.023 breakages per nt, biased towards G and A</li>
                                <li>10x physical coverage, 30x sequencing depth</li>
                                <li>Presence of a random CT-tail at the end of sequencing reads from sequencing preparation</li>
							</ul>
						</div>
					  </div>
					</div>
					<div class="accordion-item">
						<h2 class="accordion-header" id="flush-headingThree">
						  <button class="accordion-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseThree" aria-expanded="false" aria-controls="flush-collapseThree">
							Photolithographic DNA synthesis
						  </button>
						</h2>
						<div id="flush-collapseThree" class="accordion-collapse collapse" aria-labelledby="flush-collapseThree" data-bs-parent="#accordionFlushExample">
						  <div class="accordion-body">
							Application of photolithographic synthesis in a DNA-of-things context, assuming errors representative of photolithographic synthesis with a high physical coverage and sequencing depth. In addition, the beginning and end of each sequence are randomly truncated.
							<ul class="pt-3">
								<li>0.075 deletions per nt, 0.012 insertions per nt, and 0.025 substitutions per nt</li>
								<li>200x physical coverage, 50x sequencing depth</li>
								<li>Both oligo ends randomly truncated</li>
							</ul>
						  </div>
						</div>
					  </div>
				  </div>

			</div>
		</div>

		<div class="col-md-5 col-lg-4">
			<div class="mb-2">
				<h4 class="mb-0">
					<span class="">Error profile</span><br>
				</h4>
				<span><small class="text-muted fw-light">Estimated from challenge</small></span>
			</div>
			<ul class="list-group mb-3">
				<li class="list-group-item d-flex justify-content-between lh-sm">
					<div>
						<h6 class="my-0">Substitutions</h6>
						<small class="text-muted fw-light">overall per nt</small>
					</div>
					<span class="text-muted align-self-center" id="errorprofile-substitutions">1.2%</span>
				</li>
				<li class="list-group-item d-flex justify-content-between lh-sm">
					<div>
						<h6 class="my-0">Deletions</h6>
						<small class="text-muted fw-light">overall per nt</small>
					</div>
					<span class="text-muted align-self-center" id="errorprofile-deletions">1.2%</span>
				</li>
				<li class="list-group-item d-flex justify-content-between lh-sm">
					<div>
						<h6 class="my-0">Insertions</h6>
						<small class="text-muted fw-light">overall per nt</small>
					</div>
					<span class="text-muted align-self-center" id="errorprofile-insertions">0.0%</span>
				</li>
				<li class="list-group-item d-flex justify-content-between lh-sm">
					<div>
						<h6 class="my-0">Coverage bias</h6>
						<small class="text-muted fw-light">overall</small>
					</div>
					<span class="text-muted align-self-center" id="errorprofile-bias">0.53</span>
				</li>
			</ul>
			<br>
			<button class="w-100 btn btn-primary btn-lg" type="submit">Submit job</button>

		</div>
	</div>
</form>


{% endblock content_body %}